alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,263 @@
+    This algorithm generates the intrinsic reward based on the mutual information estimation between the states of interests and the context states.
+
+    See Zhao et al "Self-Supervised State-Control through Intrinsic Mutual Information Rewards"
+    """
+
+    def __init__(self,
+                 feature_spec,
+                 traj_spec,
+                 soi_spec,
+                 soc_spec,
+                 mi_r_scale=5000,
+                 hidden_size=256,
+                 misc_layerx: Network = None,
+                 misc_layery: Network = None,
+                 misc_layero: Network = None,
+                 buffer_size=100,
+                 n_objects=1,
+                 empowerment=False,
+                 env_name='SocialBot-PlayGround-v0',
+                 name="MISCAlgorithm"):
+        """Create an MISCAlgorithm.
+        
+        Args:
+            feature_spec: size of the state and the action
+            traj_spec: (num_parallel_environments, feature_spec)
+            soi_spec: (None, None, state of interest size)
+            soc_spec: (None, None, state of context size)
+            mi_r_scale: scale factor of mi estimation
+            hidden_size: number of hidden units in neural nets
+            misc_layerx: input layer with context states as input, output dimension is of hidden_size
+            misc_layery: input layer with states of interest as input, output dimension is of hidden_size
+            misc_layero: output layer with with input shape (None, None, hidden_size), output dimension is of 1
+            buffer_size: buffer size for the data buffer storing the trajectories for training the Mutual Information Neural Estimator
+            n_objects: number of objects for estimating the mutual information
+            empowerment: if empowerment is true, then use action instead of the context states
+            name: the algorithm name, "MISCAlgorithm"
+        """
+
+        super(MISCAlgorithm, self).__init__(
+            train_state_spec=feature_spec, name=name)
+
+        feature_dim = tf.nest.flatten(feature_spec)[0].shape[-1]
+
+        if misc_layerx is None:
+            misc_layerx = tf.keras.layers.Dense(
+                hidden_size, input_shape=soc_spec.shape, activation=None)
+            misc_layery = tf.keras.layers.Dense(
+                hidden_size, input_shape=soi_spec.shape, activation=None)
+            misc_layero = tf.keras.layers.Dense(
+                1, input_shape=(None, None, hidden_size), activation=None)
+
+        self._misc_layerx = misc_layerx
+        self._misc_layery = misc_layery
+        self._misc_layero = misc_layero
+
+        self._traj_spec = traj_spec
+        self._buffer_size = buffer_size
+        self._buffer = DataBuffer(self._traj_spec, capacity=self._buffer_size)
+        self._mi_r_scale = mi_r_scale
+        self._n_objects = n_objects
+        self._empowerment = empowerment
+        self._env_name = env_name
+
+    def split_observation_tf(self, o):
```

Comment:
This function REALLY needs to be provided by the user of this class as it is task dependent. You can add an argument for `__init__` such as `split_observation_func` which accepts a function as its value.

Reply:
It has already been changed see __init__().

Reply:
separate split function as required